specs/release.md

```diff
@@ -0,0 +1,319 @@
+
+## Continuous Integration Pipeline
+
+Every push to any branch and every pull request MUST trigger automated CI checks that validate code quality, correctness, and test coverage.
+
+### CI Workflow Stages
+
+The CI pipeline MUST execute the following stages in order, failing fast at the first error:
+
+**1. Type Checking:** TypeScript compilation MUST succeed with no type errors. The project uses strict mode and all type checking rules MUST be enforced.
+
+**2. Unit Tests:** All unit tests MUST pass with deterministic results. Flaky tests MUST be fixed or quarantined. The test suite MUST complete in under 60 seconds to provide rapid feedback.
+
+**3. E2E Tests:** End-to-end integration tests MUST pass, validating the full OAuth 2.1 + PKCE + MCP protocol flow. E2E tests MUST complete in under 30 seconds and MUST NOT require manual intervention or browser interaction.
+
+**4. Coverage Validation:** Code coverage MUST meet minimum thresholds (95% statement coverage, 95% function coverage). Coverage regressions MUST fail the build.
+
+### CI Execution Environment
+
+CI jobs MUST run on GitHub Actions using Ubuntu latest runners. The environment MUST use:
+
+- Node.js 20 (LTS)
+- pnpm 9.0.0 (via corepack)
+- Frozen lockfile (`pnpm install --frozen-lockfile`) to ensure reproducible builds
+
+CI jobs MUST use cached dependencies to reduce execution time. Cache invalidation MUST occur on `pnpm-lock.yaml` changes.
+
+## Continuous Deployment Pipeline
+
+Deployments are triggered automatically based on branch and merge events, with different environments receiving deployments under different conditions.
+
+### Development Deployment
+
+Every commit merged to `main` MUST automatically deploy to the development environment. This deployment occurs after all CI checks pass and MUST NOT require manual approval.
+
+The development deployment workflow MUST:
+
+1. Verify all CI checks passed (tests, type checking, coverage)
+2. Deploy to Cloudflare Workers development environment
+3. Create a GitHub Deployment record with environment "development"
+4. Update deployment status to "success" or "failure"
+5. Verify deployment with health check endpoint
+6. Rollback automatically if health check fails
+
+Development deployments MUST complete within 3 minutes of merge to `main`. Deployment failures MUST NOT block subsequent development deployments (fail forward, not fail stop).
+
+### Production Deployment
+
+Production deployments MUST be triggered manually via GitHub Actions workflow dispatch. This manual gate provides control over production releases while maintaining automation for everything except the final approval decision.
+
+The production deployment workflow MUST:
+
+1. Accept a workflow dispatch event from an authorized maintainer
+2. Verify all CI checks passed on the commit being deployed
+3. Re-run tests as final validation (even if CI already passed)
```

Comment:
I think if we can genuinely verify that this commit has already passed all of this testing with no further modifications then I don't think there's value in running them again. If we do even the tiniest bit of house keeping though and there ends up being a new commit(version bump, changelog, etc) then we should rerun.